### PR TITLE
[Backward Incompatible] Multi cluster refactor and abort execution fix

### PR DESCRIPTION
**Pull request by anandswaminathan**
This PR has the following changes

- The cluster config section and is backward incompatible. Removes currentCluster from the config
- Refactored the config and client setup to support Admin talking to Multiple clusters.
- Have a new config called as Mode, that determines what cluster selector to use.
- Stores the execution cluster info in the execution model to fix abort issues on failover.

Updated the cluster resource to make sure that resource creation happens on the correct target.

Should be fairly easy to get rid of config and inject a database for clusters as needed.

### CODE REVIEW

**Comment by anandswaminathan** [comment]
cc @katrogan @surindersinghp @akhurana001 @jonathanburns

File: pkg/executioncluster/random_cluster_selector_test.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package executioncluster
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func getRandomeMultiClusterForTest() RandomClusterSelector {
```
Comment (katrogan) [inline pkg/executioncluster/random_cluster_selector_test.go]: nit: getRandomeMultiClusterForTest -> getRandomMultiClusterForTest

File: pkg/flytek8s/client.go [hunk 2]
```diff
@@ -66,9 +70,9 @@ func GetRestClientConfig(kubeConfig, master string,
 	return kubeConfiguration, nil
 }
 
-// Initializes a kubernetes Client which performs CRUD operations on Kubernetes objects
-func NewKubeClient(kubeConfig, master string, configuration runtimeInterfaces.ClusterConfiguration) (client.Client, error) {
-	kubeConfiguration, err := GetRestClientConfig(kubeConfig, master, configuration)
+// Iniitalizes a kubernetes Client which performs CRUD operations on Kubernetes objects
```
Comment (katrogan) [inline pkg/flytek8s/client.go]: nit: revert misspelling here?

File: pkg/manager/impl/execution_manager.go [hunk 3]
```diff
@@ -677,7 +677,7 @@ func (m *ExecutionManager) publishNotifications(ctx context.Context, request adm
 	logger.Debugf(ctx, "publishing notifications for execution [%+v] in state [%+v] for notifications [%+v]",
 		request.Event.ExecutionId, request.Event.Phase, notificationsList)
 	for _, notification := range notificationsList {
-		// Check if the notification phase matches the current one.
+		// 1. Check if the notification phase matches the current one.
```
Comment (katrogan) [inline pkg/manager/impl/execution_manager.go]: revert? there aren't any other numbers below

File: pkg/manager/impl/execution_manager.go [hunk 4]
```diff
@@ -743,6 +738,15 @@ func (m *ExecutionManager) TerminateExecution(
 		logger.Infof(ctx, "couldn't find execution [%+v] to save termination cause", request.Id)
 		return nil, err
 	}
+
+	err = m.workflowExecutor.TerminateWorkflowExecution(ctx, workflowengineInterfaces.TerminateWorkflowInput{
```
Comment (katrogan) [inline pkg/manager/impl/execution_manager.go]: is this intentional? by design we've intentionally saved in the db only after we successfully deleted the workflow in kubernetes

File: pkg/runtime/interfaces/cluster_configuration.go [hunk 5]
```diff
@@ -19,6 +20,12 @@ type Auth struct {
 	CertPath  string `json:"certPath"`
 }
 
+type ClusterSelection string
```
Comment (katrogan) [inline pkg/runtime/interfaces/cluster_configuration.go]: nit: maybe ClusterSelectionStrategy or something more descriptive?

File: pkg/workflowengine/impl/propeller_executor.go [hunk 6]
```diff
@@ -87,64 +84,81 @@ func (c *FlytePropeller) addPermissions(launchPlan admin.LaunchPlan, flyteWf *v1
...
 
-	labels := addMapValues(inputs.Labels, flyteWf.Labels)
+	labels := addMapValues(input.Labels, flyteWf.Labels)
 	flyteWf.Labels = labels
-	annotations := addMapValues(inputs.Annotations, flyteWf.Annotations)
+	annotations := addMapValues(input.Annotations, flyteWf.Annotations)
 	flyteWf.Annotations = annotations
 
-	_, err = c.client.FlyteworkflowV1alpha1().FlyteWorkflows(namespace).Create(flyteWf)
+	targetCluster, err := c.executionCluster.GetTarget(nil)
```
Comment (katrogan) [inline pkg/workflowengine/impl/propeller_executor.go]: nit: add a comment or named variable to clarify what nil is here?

File: pkg/manager/impl/execution_manager.go (hunk 4 above)
Comment (anandswaminathan) [reply pkg/manager/impl/execution_manager.go]: Yes. It's the same.

I am only doing a Get before termination, as I need the cluster where the execution was kicked off

File: pkg/executioncluster/execution_target.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"github.com/lyft/flyteadmin/pkg/flytek8s"
+	runtime "github.com/lyft/flyteadmin/pkg/runtime/interfaces"
+	flyteclient "github.com/lyft/flytepropeller/pkg/client/clientset/versioned"
+	"github.com/lyft/flytestdlib/promutils"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ExecutionTargetSpec struct {
```
[the remark this answers is not in the document]
Comment (anandswaminathan) [reply pkg/executioncluster/execution_target.go]: Fixed.

File: pkg/executioncluster/random_multi_cluster.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func (s RandomMultiCluster) GetTarget(spec *ExecutionTargetSpec) (*ExecutionTarget, error) {
+	if spec != nil {
+		if val, ok := s.executionTargetMap[spec.TargetId]; ok {
+			return &val, nil
+		}
+		return nil, errors.New(fmt.Sprintf("invalid cluster target %s", spec.TargetId))
+	}
+	len := len(s.executionTargetMap)
```
[the remark this answers is not in the document]
Comment (anandswaminathan) [reply pkg/executioncluster/random_multi_cluster.go]: already resolved in refactor

File: pkg/flytek8s/client.go [hunk 9]
```diff
@@ -66,9 +70,9 @@ func GetRestClientConfig(kubeConfig, master string,
 	return kubeConfiguration, nil
 }
 
-// Initializes a kubernetes Client which performs CRUD operations on Kubernetes objects
-func NewKubeClient(kubeConfig, master string, configuration runtimeInterfaces.ClusterConfiguration) (client.Client, error) {
-	kubeConfiguration, err := GetRestClientConfig(kubeConfig, master, configuration)
+// Iniitalizes a kubernetes Client which performs CRD operations on Kubernetes objects
```
Comment (katrogan) [inline pkg/flytek8s/client.go]: please don't introduce misspellings! (Iniitalizes)

**Pull request merged**